ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java

```diff
@@ -0,0 +1,411 @@
+  private Handler<AsyncResult<Void>> limitHandler = ar -> { };
+  private Runnable queueCallback = () -> { };
+
+  public PullQueryWriteStream(
+      final OptionalInt hardLimit,
+      final StreamedRowTranslator translator
+  ) {
+    this.hardLimit = hardLimit;
+    this.translator = translator;
+
+    // register a drainHandler that will wake up anyone waiting on hasCapacity
+    drainHandler.add(ignored -> {
+      monitor.enter();
+      monitor.leave();
+    });
+  }
+
+  private static final class HandledRow {
+    private final PullQueryRow row;
+    private final Handler<AsyncResult<Void>> handler;
+
+    private HandledRow(
+        final PullQueryRow row,
+        final Handler<AsyncResult<Void>> handler
+    ) {
+      this.row = row;
+      this.handler = handler;
+    }
+  }
+
+  // -------------------- READ QUEUE METHODS -------------------------
+
+  @Override
+  public void drainTo(final Collection<? super KeyValueMetadata<List<?>, GenericRow>> collection) {
+    monitor.enter();
+    try {
+      while (!queue.isEmpty()) {
+        collection.add(poll());
+      }
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  public void drainRowsTo(final Collection<PullQueryRow> collection) {
+    monitor.enter();
+    try {
+      while (!queue.isEmpty()) {
+        collection.add(pollRow());
+      }
+    } finally {
+      monitor.leave();
+    }
+  }
+
+  @Override
+  public KeyValueMetadata<List<?>, GenericRow> poll(
+      final long timeout,
+      final TimeUnit unit
+  ) throws InterruptedException {
+    if (monitor.enterWhen(hasData, timeout, unit)) {
+      try {
+        return poll();
+      } finally {
+        monitor.leave();
+      }
+    }
+
+    return null;
+  }
+
+  @Override
+  public KeyValueMetadata<List<?>, GenericRow> poll() {
+    final PullQueryRow row = pollRow();
+    if (row == null) {
+      return null;
+    }
+
+    if (row.getConsistencyOffsetVector().isPresent()) {
+      return new KeyValueMetadata<>(RowMetadata.of(row.getConsistencyOffsetVector().get()));
+    }
+
+    return new KeyValueMetadata<>(
+        KeyValue.keyValue(null, row.getGenericRow()),
+        row.getSourceNode()
+            .map(sn -> new KsqlHostInfo(sn.getHost(), sn.getPort()))
+            .map(RowMetadata::of)
+    );
+  }
+
+  public PullQueryRow pollRow(
+      final long timeout,
+      final TimeUnit unit
+  ) throws InterruptedException {
+    if (monitor.enterWhen(hasData, timeout, unit)) {
+      try {
+        return pollRow();
+      } finally {
+        monitor.leave();
+      }
+    }
+
+    return null;
+  }
+
+  private PullQueryRow pollRow() {
+    final HandledRow polled;
+    monitor.enter();
+    try {
+      polled = queue.poll();
+    } finally {
+      monitor.leave();
+    }
+
+    if (polled == null) {
+      return null;
+    }
+
+    polled.handler.handle(FUTURE_FACTORY.succeededFuture());
+
+    monitor.enter();
+    try {
+      if (queue.size() < softLimit) {
```

Comment:
Is the drain handler meant to be called once the write stream is entirely drained, or when it has additional capacity?  Works either way, but just curious about the API.  By the comment here, I think you're right: https://vertx.io/docs/apidocs/io/vertx/core/streams/WriteStream.html#drainHandler-io.vertx.core.Handler-

Reply:
just when it has additional capacity... it seems like a bad name

Reply:
Actually looks like it's up to us: 
> The stream implementation defines when the drain handler, for example it could be when the queue size has been reduced to maxSize / 2.

maybe I'll change it to that